More PICA200 Emulation Fixes

neobrain opened the pull request:
Touches lots of stuff in basically all areas of the PICA200 code. No real game changers other than rounding off existing code by bugfixing and/or implementing particular configurations.

yuriks commented on src/core/hle/service/gsp_gpu.h:
```diff
@@ -112,6 +112,10 @@ struct Command {
+            union {
+                BitField< 0, 16, u32> control1;
+                BitField<16, 16, u32> control2;
+            };
```
Why not simply use two `u16`s here?

Subv replied:
I agree with @yuriks

bunnei replied:
+1

neobrain replied:
ack.

neobrain commented:
Pending on #561, btw, but that's not relevant for review and an easy rebase later on.

yuriks commented on src/core/hw/gpu.cpp:
```diff
@@ -68,22 +68,37 @@ inline void Write(u32 addr, const T data) {
-        const bool is_second_filler = (index != GPU_REG_INDEX(memory_fill_config[0].value));
-        const auto& config = g_regs.memory_fill_config[is_second_filler];
+        const bool is_second_filler = (index != GPU_REG_INDEX(memory_fill_config[0].trigger));
+        auto config = g_regs.memory_fill_config[is_second_filler];
```
You dropped the `&` here, so the sets at the end of the fill won't be doing anything.

neobrain replied:
Good catch, not sure why I dropped the `&`.

yuriks commented on src/core/hw/gpu.cpp:
```diff
@@ -68,22 +68,37 @@ inline void Write(u32 addr, const T data) {
 
     // Memory fills are triggered once the fill value is written.
     // NOTE: This is not verified.
```
This comment seems inadequate now.

yuriks commented on src/core/hw/gpu.cpp:
```diff
@@ -68,22 +68,37 @@ inline void Write(u32 addr, const T data) {
+                    ptr[2] = config.value_24bit_r;
+                }
+            } else if (config.fill_32bit) {
+                for (u32* ptr = (u32*)start; ptr != (u32*)end; ++ptr)
```
Is this actually `!=` and not `<`? (Same comment for 16-bit)

kevinhartman replied:
!= should be also be <

Alegend45 commented:
Other than yuriks' comments, LGTM

yuriks commented on src/core/hw/gpu.cpp:
```diff
@@ -68,22 +68,37 @@ inline void Write(u32 addr, const T data) {
+            u8* start = Memory::GetPointer(Memory::PhysicalToVirtualAddress(config.GetStartAddress()));
+            u8* end = Memory::GetPointer(Memory::PhysicalToVirtualAddress(config.GetEndAddress()));
+
+            if (config.fill_24bit) {
```
I have a feeling these loops will generate dreadful code due to acessing the structs directly. I'd recommend copying the fill values to local variables after checking the assembly.

neobrain replied:
What would be the issue with "accessing the structs directly"? I don't see any difference it would make compared to reading any value stored in RAM.

yuriks replied:
The fill source is constant, so not constantly reloading it from memory (Which the compiler is more likely to do if you're accessing it through a struct and functions that run code.) and instead keeping it on registers would produce a faster loop.

neobrain replied:
hmm I guess that's something to worry about when someone with time on his hands pins down a bottleneck here due to what you explained. For now, it's just redundant code for little reason, and I don't really have the time or interest to start analyzing disassembly code for this.

yuriks replied:
*sigh*

lioncash commented on src/video_core/command_processor.cpp:
```diff
@@ -71,6 +69,11 @@ static inline void WritePicaReg(u32 id, u32 value, u32 mask) {
+            // HACK: Initialize vertex_attribute_elements to zero to prevent infinite loops below.
+            // This is one of the hacks required to deal with uninitalized vertex attributes.
+            // TODO: Fix this properly.
+            boost::fill(vertex_attribute_elements, 0);
```
You can zero-fill the array just by adding `= {};` to the end of the declaration.

So `u32 vertex_attribute_elements[16] = {};`

neobrain replied:
Fixed, thanks :)

kevinhartman commented on src/core/hw/gpu.cpp:
```diff
@@ -68,22 +68,37 @@ inline void Write(u32 addr, const T data) {
+                    *ptr = config.value_32bit;
+            } else {
+                // 16-bit fill
+                for (u16* ptr = (u16*)start; ptr != (u16*)end; ++ptr)
```
!= should also be <

yuriks commented on src/video_core/clipper.cpp:
```diff
@@ -117,14 +87,31 @@ void ProcessTriangle(OutputVertex &v0, OutputVertex &v1, OutputVertex &v2) {
+    static const float24 zero = float24::FromFloat32(0.0);
+    static const float24 one = float24::FromFloat32(1.0);
+    static const std::array<ClippingEdge, 7> clipping_edges = {{
+        { Math::MakeVec(one, zero, zero, -one) },   // x = +w
```
I think this table would be easier to read if you named the variables `f0` and `f1` (or similar).

neobrain replied:
Indeed.

yuriks commented on src/video_core/clipper.cpp:
```diff
@@ -117,14 +87,31 @@ void ProcessTriangle(OutputVertex &v0, OutputVertex &v1, OutputVertex &v2) {
+    // TODO: Test if dropping the whole primitive in this case reflects hardware behavior.
+/*    ClippingEdge near_edge{ ClippingEdge::POS_Z, Math::Vec4<float24>(float24::FromFloat32(0), float24::FromFloat32(0), float24::FromFloat32(1), float24::FromFloat32(0)) };
+    if (near_edge.IsOutSide(v0) || near_edge.IsOutSide(v1) || near_edge.IsOutSide(v2))
+        return;*/
```
What's this about?

neobrain replied:
At least some platforms (GC+Wii) drop whole primitives as soon as one vertex lies outside the near z-clipping plane, IIRC for optimization purposes. I'll clean this line up, though.

neobrain replied:
Actually, I'll drop the code sample altogether and just elaborate the TODO instead.

yuriks commented on src/video_core/clipper.cpp:
```diff
@@ -117,14 +87,31 @@ void ProcessTriangle(OutputVertex &v0, OutputVertex &v1, OutputVertex &v2) {
+    if (near_edge.IsOutSide(v0) || near_edge.IsOutSide(v1) || near_edge.IsOutSide(v2))
+        return;*/
+
+    // NOTE: We clip against a w=epsilon plane to guarantee that the output has a positive w value.
```
Why not simply "> 0.f"?

neobrain replied:
How would I implement that condition? It would end up with lots of vertices **in** the w=0 plane, which will divide by zero in the perspective divide.

yuriks replied:
My point is that you can set the plane to 0 if you discard the vertices *in* the plane, by passing when `w > 0.f`, instead of `w >= 0.f`, which you already seem to be doing anyway in L26.

yuriks replied:
Disregard this.

yuriks commented on src/video_core/rasterizer.cpp:
```diff
@@ -20,16 +21,31 @@ namespace Rasterizer {
+
+    default:
+        LOG_CRITICAL(Render_Software, "Unknown framebuffer color format %x", registers.framebuffer.color_format);
+        exit(1);
```
Make this an invocation of the `UNIMPLEMENTED` macro instead.

yuriks commented on src/video_core/rasterizer.cpp:
```diff
@@ -612,79 +676,70 @@ void ProcessTriangle(const VertexShader::OutputVertex& v0,
+
+                    default:
+                        LOG_CRITICAL(HW_GPU, "Unknown RGB blend equation %x", equation);
+                        exit(0);
```
Use `UNIMPLEMENTED()`

yuriks commented on src/video_core/rasterizer.cpp:
```diff
@@ -612,79 +676,70 @@ void ProcessTriangle(const VertexShader::OutputVertex& v0,
+                blend_output.a() = EvaluateBlendEquation(combiner_output, srcfactor, dest, dstfactor, params.blend_equation_a).a();
             } else {
                 LOG_CRITICAL(HW_GPU, "logic op: %x", registers.output_merger.logic_op);
                 exit(0);
```
`UNIMPLEMENTED()`

yuriks commented on src/video_core/vertex_shader.cpp:
```diff
@@ -85,8 +85,11 @@ struct VertexShaderState {
+        u32 return_address; // Where to jump when leaving scope
+        u8 repeat_counter;  // How often to repeat until this call stack element is removed
+        u8 loop_increment;  // Which value to add to the loop counter after an iteration
+                            // TODO: Should this be a signed value? Does it even matter?
```
Probably shouldn't be a signed, that'll only invite more C UB

bunnei commented:
Looks like you made a typo somewhere!

![](http://i.imgur.com/uclCDZl.png)

yuriks commented:
Man, that looks awesome! :D Looks like an issue with wrong strides.

Alegend45 commented:
Aaaand boom goes the dynamite.

bunnei commented on src/video_core/clipper.cpp:
```diff
@@ -15,30 +15,15 @@ namespace Clipper {
-
-    ClippingEdge(Type type, float24 position) : type(type), pos(position) {}
+    ClippingEdge(Math::Vec4<float24> coeffs,
+                 Math::Vec4<float24> bias = Math::Vec4<float24>(float24::FromFloat32(0), float24::FromFloat32(0), float24::FromFloat32(0), float24::FromFloat32(0)))
```
Very long line

bunnei commented on src/video_core/rasterizer.cpp:
```diff
@@ -230,12 +254,24 @@ void ProcessTriangle(const VertexShader::OutputVertex& v0,
 
+                        case Regs::TextureConfig::MirroredRepeat:
+                        {
+                            int val = (int)(((unsigned)val) % (2*size));
```
spaces please, `(2 * size)`

bunnei commented on src/video_core/rasterizer.cpp:
```diff
@@ -230,12 +254,24 @@ void ProcessTriangle(const VertexShader::OutputVertex& v0,
                 };
+
+                // Textures are laid out from bottom to top, hence we invert the t coordinate.
+                // NOTE: This may not be the right to place the inversion.
```
Typo: `// NOTE: This may not be the right place for the inversion.`

bunnei commented:
Read through the PR. Don't have much to say in addition to what was already said, as well as to fix the texturing regression from the previous pica_progress3.

Alegend45 commented:
I'm thinking this might be the bad commit: https://github.com/neobrain/citra/commit/f4e4ef981406bab1d91ec1dd0a5f5b6318949b5d

neobrain commented:
@Alegend45 Not sure how you tested things, but neobrain@92a1c9d4666c19 sounds more likely to be the culprit. My local changes have a fix for that, which I forgot to include in this PR.

neobrain commented:
Addressed review comments and fixed that bug. OoT3D will still need the culling hack for things to show up at all, and I'm reasonably sure that issue is not actually related to the backface culling code.

In other words: Ready for merging.

neobrain commented:
Addressed review comments and fixed that bug. OoT3D will still need the culling hack for things to show up at all, and I'm reasonably sure that issue is not actually related to the backface culling code.

In other words: Ready for merging.

yuriks commented:
Looks good to merge.

bunnei commented:
LGTM
